framework/src/main/java/org/checkerframework/framework/StubTypes.java

```diff
@@ -0,0 +1,476 @@
+     */
+    public AnnotatedTypeMirror getAnnotatedTypeMirror(Element e) {
+        if (parsing) {
+            return null;
+        }
+        parseEnclosingClass(e);
+        AnnotatedTypeMirror type = typesFromStubFiles.get(e);
+        return type == null ? null : type.deepCopy();
+    }
+
+    /**
+     * Returns the set of declaration annotations for {@code e} containing only annotations
+     * explicitly written in a stub file or the empty set if {@code e} does not appear in a stub
+     * file.
+     *
+     * @param elt element for which annotations are returned
+     * @param eltName name of the element
+     * @return an AnnotatedTypeMirror for {@code e} containing only annotations explicitly written
+     *     in the stubfile and in the element. {@code null} is returned if {@code element} does not
+     *     appear in a stub file.
+     */
+    public Set<AnnotationMirror> getDeclAnnotation(Element elt, String eltName) {
+        if (parsing) {
+            return Collections.emptySet();
+        }
+
+        parseEnclosingClass(elt);
+
+        if (declAnnosFromStubFiles.containsKey(eltName)) {
+            return declAnnosFromStubFiles.get(eltName);
+        }
+        return Collections.emptySet();
+    }
+
+    /**
+     * Parses the outermost enclosing class of {@code e} if there exists a stub file for it and it
+     * has not already been parsed.
+     */
+    private void parseEnclosingClass(Element e) {
+        String className = getOuterMostEnclosingClass(e);
+        if (className == null) {
+            return;
+        }
+        if (jdk11StubFiles.containsKey(className)) {
+            parseStubFile(jdk11StubFiles.get(className));
+            jdk11StubFiles.remove(className);
+        } else if (jdk11StubFilesJar.containsKey(className)) {
+            parseJarEntry(jdk11StubFilesJar.get(className));
+            jdk11StubFilesJar.remove(className);
+        }
+    }
+
+    /**
+     * @return the fully qualified name of the outermost enclosing class of {@code e} or {@code
+     *     null} if no such class exists for {@code e}.
+     */
+    private String getOuterMostEnclosingClass(Element e) {
+        TypeElement enclosingClass = ElementUtils.enclosingClass(e);
+        if (enclosingClass == null) {
+            return null;
+        }
+        while (true) {
+            Element element = enclosingClass.getEnclosingElement();
+            if (element == null || element.getKind() == ElementKind.PACKAGE) {
+                break;
+            }
+            TypeElement t = ElementUtils.enclosingClass(element);
+            if (t == null) {
+                break;
+            }
+            enclosingClass = t;
+        }
+        return enclosingClass.getQualifiedName().toString();
+    }
+
+    /**
+     * Parses the stub file in {@code path}.
+     *
+     * @param path path to file to parse
+     */
+    private void parseStubFile(Path path) {
+        boolean oldParsing = parsing;
```

Comment:
It isn't clear why the state of `parsing` needs to be handled like this. Can you add a comment on the field? Or here?
I would have expected that if parsing is already true, you raise an error.